### PR TITLE
feat(*): misc improvements for Sequelize v5

### PR DESCRIPTION
**Pull request by SimonSchick**
Some things:
- NodeJS 10 build target
- Removed all usage of `$operator` with symbol equivalents (requires TS 2.7)
- Moved core hook methods into it's own class, typed generic hook functions and added missing hooks
- Added ValidationOptions interface and replaced many instances of `object`.
- Misc typings improvements for `Transaction` (uses namespace, **please advise**)
- Added `config` property to `Sequelize`
- Improvements typings for `where` `literal` etc. static functions
- Removed some redundant JSDoc types.
- Updated validator typings
- Changes lib to es2016 **please advise**

- Some tests for changes related above

### CODE REVIEW

**Comment by SimonSchick** [comment]
@felixfbecker could you take a look?

File: lib/deferrable.d.ts [hunk 1]
```diff
@@ -34,10 +34,10 @@
  *
  */
 export interface AbstractDeferrableStatic {
-    new (): AbstractDeferrable
-    (): AbstractDeferrable
+    new (): Deferrable
+    (): Deferrable
 }
-export interface AbstractDeferrable {
+export interface Deferrable {
```
Comment (SimonSchick) [inline lib/deferrable.d.ts]: Note: Although supported, shouldn't we make this a class and remove the call without `new`?

File: lib/model.d.ts [hunk 2]
```diff
@@ -107,17 +109,17 @@ export interface ScopeOptions {
...
 
 /**
- * Example: `$any: [2,3]` becomes `ANY ARRAY[2, 3]::INTEGER`
+ * Example: `[Op.any]: [2,3]` becomes `ANY ARRAY[2, 3]::INTEGER`
  *
  * _PG only_
  */
 export interface AnyOperator {
-    $any: (string | number)[]
+    [Op.any]: (string | number)[]
```
Comment (felixfbecker) [inline lib/model.d.ts]: This is a breaking change and should probably not be merged till v5 is out

File: lib/sequelize.d.ts [hunk 3]
```diff
@@ -144,6 +154,33 @@ export interface OperatorsAliases {
     [K: string]: symbol
 }
 
+/**
+ * Final config options generated by sequelize.
+ */
+export interface Config {
+    readonly database: string
```
Comment (felixfbecker) [inline lib/sequelize.d.ts]: Why `readonly`?

File: lib/transaction.d.ts [hunk 4]
```diff
@@ -1,25 +1,50 @@
...
- * The transaction object is used to identify a running transaction. It is created by calling
- * `Sequelize.transaction()`.
- *
- * To run a query under a transaction, you should pass the transaction in the options object.
- */
-export class Transaction {
-    constructor(sequelize: Sequelize, options: TransactionOptions)
+import { Deferrable, Sequelize } from './sequelize'
 
+// tslint:disable-next-line
```
Comment (felixfbecker) [inline lib/transaction.d.ts]: please only disable specific rules and mention why they need to be disabled. Maybe the rules shouldn't be enabled in the first place.

File: lib/transaction.d.ts [hunk 5]
```diff
@@ -1,25 +1,50 @@
...
+     *
+     * }).then(result => {
+     *   // transaction has been committed. Do something after the commit if required.
+     * }).catch(err => {
+     *   // do something with the err.
+     * });
+     * ```
      */
-    public commit(): Promise<void>
+    const enum ISOLATION_LEVELS {
```
Comment (felixfbecker) [inline lib/transaction.d.ts]: imo these should just be exported top-level, which is what we do in other files if Sequelize provides both.

File: test/models/User.ts [hunk 6]
```diff
@@ -39,10 +39,14 @@ User.init(
 )
 
 // Hooks
-User.afterFind((users, options: FindOptions) => {
+User.afterFind((users, options) => {
     console.log('found')
 })
 
+// TODO: VSCode shows the typing being correctly narrowed but doesn't do it correctly
```
Comment (felixfbecker) [inline test/models/User.ts]: Why is this vscode specific? vscode uses the TS language service

File: lib/hooks.d.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+export interface SequelizeHooks extends AllModelHooks {
+    beforeConnect(config: Config): HookReturn
+    afterConnect(connection: any, config: Config): HookReturn
+}
+
+/**
+ * Virtual class for deduplication
+ */
+export class Hooks {
```
[the remark this answers is not in the document]
Comment (SimonSchick) [reply lib/hooks.d.ts]: That is correct, hence the comment, I tried to find a way to dedupe the hooks, they are basically a mixin, I wanted to first use declaration merging via interfaces but that doesn't work for statics.
Do you have any alternative suggestions?

File: lib/model.d.ts (hunk 2 above)
Comment (SimonSchick) [reply lib/model.d.ts]: Correct, this PR is meant to be for v5.

File: lib/sequelize.d.ts (hunk 3 above)
Comment (SimonSchick) [reply lib/sequelize.d.ts]: I'm pretty sure that changing this at runtime causes all sorts of issues, looking at the source code this looks like semi internal, despite not being _ prefixed.

File: test/models/User.ts (hunk 6 above)
Comment (SimonSchick) [reply test/models/User.ts]: I am not sure what is happening either, when I hover over the declaration in VSCode it correctly determines the callback type, but the text pane can't. Can you maybe repro this?

File: lib/transaction.d.ts (hunk 5 above)
Comment (SimonSchick) [reply lib/transaction.d.ts]: What do you mean with `top-level`, can you provide an example?

File: lib/deferrable.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (SimonSchick) [reply lib/deferrable.d.ts]: Calling an interface `Abstract` seems like an anti pattern just as much as using the word `Abstract` in class names, by nature interfaces are abstractions so this felt redundant.
If you want I can change it back.

**Comment by SimonSchick** [comment]
@felixfbecker sequelize v5 is live.

**Comment by SimonSchick** [comment]
Closing since there is no activity, I've forked the typings and will try to get them into DT with version 5.

**Pull request closed without merging**